### ISSUE TITLE
`System.TypedReference` should be treated as managed

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#33965]
Even once the [corefx issue](https://github.com/dotnet/corefx/issues/35873) (adding a dummy `object` field to the ref assembly for `TypedReference`) is fixed, we're not sure that the compiler will honor it properly. We may have to add a special case to recognize `TypedReference`.

The goal is that `var x = stackalloc TypedReference[10];` should be an error (because `TypedReference` is managed).

More [context](https://github.com/dotnet/roslyn/pull/33951#issuecomment-470778517).

**Comment by benaadams** [comment]
Out of curiosity why are arrays of managed types disallowed from stackalloc? They can certainly have references that live on the stack
```csharp
Span<string> strings = stackalloc string[count];
```
Would be kinda handy

**Comment by jcouv** [comment]
@benaadams I think it is a GC tracking issue. 
I don't know if there is a fundamental reason why references insides stackallocs couldn't also be tracked.

**Comment by benaadams** [comment]
@jcouv thanks, so should raise in coreclr and if supportable have it exposed as a runtime feature flag? (so C# can enquire if the targeted runtime/TFM supports it)

**Comment by jcouv** [comment]
Correct.
I think @jaredpar had discussions around that (GC tracking of stackallocs). He may be able to summarize or share a pointer.

**Comment by jaredpar** [comment]
I don't think we should be doing anything special here in the compiler. The compiler is behaving correctly according to the spec and rules. The bug is that the CoreFX team has shipped incorrect reference assemblies. This has come up several times in the past, even for this exact type of problem, and the solution is that CoreFX fixes the assemblies and the compiler does not special case types. 

The reason for this is that the customer break occurs in the correct place. Moving to a new version of the compiler is not a breaking change but adopting the new CoreFX assemblies potentially is. This is how we prefer to push CoreFX breaks to customers.

**Comment by jaredpar** [comment]
> I don't know if there is a fundamental reason why references insides stackallocs couldn't also be tracked.

Artifact of history. Under the hood `localloc` returns a pointer, not a managed type, hence the GC has never tracked the contents here. 

Now that we've made it more prominent again with `Span<T>` there has been reason to revisit this. I've spoken with the runtime team a few times about this. There is nothing fundamentally wrong here. Having the GC track the result of a stack alloc is an approachable problem.

Rather than retrofitting `localloc` though we'd just introduce a new runtime intrinsic. Imagine for example:

``` csharp
class RuntimeIntrinsics { 
  Span<T> StackAlloc<T>(int count); 
}
```

> should raise in coreclr and if supportable have it exposed as a runtime feature flag?

I think it's hard to change this as a feature flag. The feature flag can't change the return of `localloc`, it will always be a pointer. Even if we immediately stash it into a `Span<T>` there would be a moment where it existed on the execution stack as an untracked pointer. Hence there will always be a moment that the GC can run, not see a root and collect the values. Seems much more approachable to have the runtime helper and then let the compiler just change the implementation of `stackalloc` to use that if available.

**Comment by benaadams** [comment]
> Even if we immediately stash it into a `Span<T>` there would be a moment where it existed on the execution stack as an untracked pointer. 

Ah! Makes sense.

Is it tracked anywhere? 

Has come up a few times; [can use an overlarge pre-sized struct](https://github.com/dotnet/corefx/blob/a3cb00f9c43a7a53934c2cd7a3366e011380184f/src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs#L1077-L1145), but zeroing costs are at the full length; so its expensive, also can't use a `fixed` buffer in the struct so you need to have multiple fields and do indexing using `Unsafe.Add` from the ref of the first field (alternative is creating a Span with a ref, but you can't do that in .NET Standard either).

**Comment by jaredpar** [comment]
@benaadams 

> Is it tracked anywhere?

Don't believe it is on CoreCLR. It's tracked as a sub-item of the `params` proposal. 

https://github.com/dotnet/csharplang/blob/master/proposals/format.md#clr-stack-allocation-helpers

**Comment by VSadov** [comment]
> Out of curiosity why are arrays of managed types disallowed from stackalloc? They can certainly have references that live on the stack

There are some differences between references in ordinary stack variables and hypothetical references inside stackalloc. 

For any position of instruction pointer, the live variables on on stack can be statically known from reachability/flow analysis. JIT/AOT can basically precompute the liveness information for every location where GC may happen (sometimes every location inside a method) and when GC scans for roots the live references are trivially and quickly reported from those maps. 

Note that the map is specific to a jitted method and does not change. No matter how many times the method was called, recursively or concurrently, etc... 

`stackalloc` with reference types would be introducing roots dynamically. It would require a good deal of additional housekeeping to track whether a stackalloc introduced new roots to stack, how many and where they went exactly. Note that the roots do not only keep the referenced objects alive, but also need to be updated when the objects move.
The tracking would be specific to an invocation frame. Next or concurrent invocation of the method must maintain a separate map.

Not allowing managed references in dynamic stack allocations makes those allocations significantly simpler to implement and to run.

**Comment by gafter** [comment]
> Even once the [corefx issue](https://github.com/dotnet/corefx/issues/35873) (adding a dummy `object` field to the ref assembly for `TypedReference`) is fixed, we're not sure that the compiler will honor it properly.

Once `corefx` is fixed, the compiler will honor it properly.
